### PR TITLE
JC19-354-Init-anime-app-table-view

### PR DESCRIPTION
**Pull request by daoduong-saritasa**
Initialize anime app with table view

### CODE REVIEW

File: apps/vanilla/src/scripts/pagination.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+import { PAGINATION_STATE, LIMIT, SORT_SETTINGS } from './variables';
+import { renderAnimeTable } from './animeTable';
+
+import { getAnime } from './anime';
+
+/** Render pagination.
+ * @param pages Number of total pages.
+ * */
+export const paginateButton = (pages: number): void => {
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:12]: You come up with a better name, like `renderPaginateButton`. Read wiki https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#naming-functions

**Review by artem-sotnichenko-saritasa — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/pages/index.css [hunk 2]
```diff
@@ -3,3 +3,27 @@
...
+
+.sort {
+  max-width: 40%;
+}
+
+/*
+.data {
+  height: 100px;
+  width: 100%;
+} */
```
Comment (TienLuu-Saritasa) [inline apps/vanilla/src/pages/index.css:29]: Please remove your comment

File: apps/vanilla/src/scripts/anime.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Anime } from '@js-camp/core/models/anime';
+import { AnimeDto } from '@js-camp/core/dtos/anime.dto';
+import { PaginationDto } from '@js-camp/core/dtos/pagination.dto';
+import { Pagination } from '@js-camp/core/models/pagination';
+import { PaginationMapper } from '@js-camp/core/mappers/pagination.mapper';
+
+import { api } from './API';
+import { SortSetting } from './interfaces';
+
+export const getAnime = async(limit: number, page: number, sortSetting: SortSetting): Promise<Pagination<Anime>> => {
```
Comment (TienLuu-Saritasa) [inline apps/vanilla/src/scripts/anime.ts:10]: > To increase the cohesion of these pagination-related properties, it would be reasonable to unify them into a single entity. Also, [we prefer avoiding ordered params](https://wiki.saritasa.rocks/frontend/languages/ts-js/best-practices#how-to-pass-arguments-to-functionsmethods).

File: libs/core/models/aired.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  /** Anime end date. */
+  public readonly end: Date;
+
+  public constructor(data: InitArgsAired) {
+    super();
+    this.end = data.end;
+    this.start = data.start;
+  }
+}
+type InitArgsAired = OmitImmerable<Aired>;
```
Comment (TienLuu-Saritasa) [inline libs/core/models/aired.ts:18]: You could create DateRange and type with it everything related to start and end dates
```
/** Period of time. */
export interface DateRange {

  /** Start of the period. */
  readonly start: Date;

  /** End of the period. */
  readonly end: Date;
}
```

**Review by TienLuu-Saritasa — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/scripts/interfaces.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+/** Sort options. */
+export interface SortOption {
+
+  /** Sort option text. */
+  text: string;
+
+  /** Sort option value. */
+  value: string;
+}
```
[the remark this answers is not in the document]
Comment (daoduong-saritasa) [reply apps/vanilla/src/scripts/interfaces.ts:19]: I only need to add readonly for SortOption right? Cause SortSetting property will be changed when choosing sort options

[the remark this answers is not in the document]
Comment (TienLuu-Saritasa) [reply apps/vanilla/src/scripts/interfaces.ts:19]: If you need to update your sort options, you can use destructuring, f.e:
```
{
  ...options,
  text: "your text"
}
```

File: libs/core/models/aired.ts (hunk 4 above)
Comment (daoduong-saritasa) [reply libs/core/models/aired.ts:18]: I think it's just naming confusion, will it acceptable if I change Aired class to DateRange, or I have to create new DateRange interface

Comment (TienLuu-Saritasa) [reply libs/core/models/aired.ts:18]: Yeah, sorry for wrong example, you could change Aired class to DateRange

File: apps/vanilla/src/scripts/anime.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import { Pagination } from '@js-camp/core/models/pagination';
+import { PaginationMapper } from '@js-camp/core/mappers/pagination.mapper';
+
+import { api } from './API';
+import { PaginationOptions } from './interfaces';
+
+export const getAnime = async({ limit, page, sortSettings }: PaginationOptions): Promise<Pagination<Anime>> => {
+  const offset: number = limit * (page - 1);
+  const { data } = await api.get<PaginationDto<AnimeDto>>(
+    `anime/anime/?limit=${limit}&offset=${offset}&ordering=${sortSettings.direction}${sortSettings.option}`,
```
Comment (TienLuu-Saritasa) [inline apps/vanilla/src/scripts/anime.ts:13]: Please check [this](https://masteringjs.io/tutorials/axios/get-query-params). The 2nd parameter to axios.get() is the [Axios options](https://github.com/axios/axios#request-config): Axios will serialize options.params and add it to the query string

File: apps/vanilla/src/scripts/pagination.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      </a>
+    </li>`;
+  }
+  wrapper.innerHTML = `
+    <li id="left" index=1 class="waves-effect paginate page"><a>First</a></li>
+    ${wrapper?.innerHTML}`;
+
+  wrapper.innerHTML += `
+  <li id="left" index=${pages} class="waves-effect paginate page"><a>Last</a></li>`;
+  const pageBtn = document.querySelectorAll('.page');
```
Comment (TienLuu-Saritasa) [inline apps/vanilla/src/scripts/pagination.ts:47]: Please [avoid contractions](https://wiki.saritasa.rocks/frontend/languages/ts-js/naming#avoid-contractions)

File: libs/core/dtos/anime.dto.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  readonly image: string;
+
+  /** Anime aired date. */
+  readonly aired: AiredDto;
+
+  /** Anime Status. */
+  readonly status: string;
+
+  /** Anime type. */
+  readonly type: 'GENRES';
```
Comment (TienLuu-Saritasa) [inline libs/core/dtos/anime.dto.ts:32]: Please create `enum` or `union`for this `type`

File: libs/core/dtos/aired.dto.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+/** Aired time DTO. */
+export interface AiredDto {
+
+  /** Aired start, for example, "2014-12-20T17:30:50.416Z". */
+  readonly start: Date;
```
Comment (TienLuu-Saritasa) [inline libs/core/dtos/aired.dto.ts:5]: > Are you sure the BE sends you `Date` object? I believe it's a `string`

**Review by timofei-maximov-saritasa — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/pages/index.css [hunk 10]
```diff
@@ -3,3 +3,21 @@
 .title {
   color: var(--heading-color-1);
 }
+
+.anime-image {
+  width: 150px;
+  height: 150px;
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/pages/index.css:9]: It can make the picture display scaling, pls check it

File: apps/vanilla/src/global/components.css [hunk 11]
```diff
@@ -1,28 +1,12 @@
...
-
 .links {
   display: flex;
   flex-wrap: wrap;
   gap: 16px;
 }
 
 .default-layout {
-  --default-layout-width: 600px;
+  --default-layout-width: 90%;
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/global/components.css:8]: It look unusual, could you move it into variables.css

File: apps/vanilla/src/scripts/API.ts [hunk 12]
```diff
@@ -0,0 +1,5 @@
+import axios from 'axios';
+export const api = axios.create({
+  baseURL: `${import.meta.env.VITE_API_URL}`,
+  headers: { 'Api-Key': `${import.meta.env.VITE_API_KEY}`, 'Accept': 'application/json' },
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/API.ts:4]: I see we have eslint warning.
We could use disable eslint
// eslint-disable-next-line @typescript-eslint/naming-convention

File: apps/vanilla/src/scripts/functions.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+/**
+ * Checks whether has sort options enabled.
+ * @param option Option value to check.
+ */
+export const hasSortOption = (option: string): boolean => option !== '';
+
+/**
+ * Change state of the order direction.
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/functions.ts:22]: I think we do not need to have Annotations on this variables, IDE is warning "it can be Null"
we can check:
 if (sortDirection === null) {
  throw new Error .... 
 }

File: apps/vanilla/src/scripts/pagination.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+import { renderAnimeTable } from './animeTable';
+
+import { getAnime } from './anime';
+import { PaginationOptions } from './interfaces';
+
+/**
+ * Render pagination.
+ * @param pages Number of total pages.
+ * */
+export const renderPaginateButton = (pages: number): void => {
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:15]: Please check 'wrapper' value can be "null"

File: apps/vanilla/src/scripts/pagination.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    </li>`;
+  }
+  wrapper.innerHTML = `
+    <li id="left" index=1 class="waves-effect paginate page"><a>First</a></li>
+    ${wrapper?.innerHTML}`;
+
+  wrapper.innerHTML += `
+  <li id="left" index=${pages} class="waves-effect paginate page"><a>Last</a></li>`;
+  const paginateButton = document.querySelectorAll('.page');
+  paginateButton.forEach(element => {
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:50]: please check
element.getAttribute('index') can be "null" value

File: apps/vanilla/src/scripts/sort.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+import { renderAnimeTable } from './animeTable';
+import { PAGINATION_STATE, SORT_SETTINGS, SORT_DIRECTION, SORT_OPTIONS, LIMIT, FIRST_PAGE } from './variables';
+import { getAnime } from './anime';
+import { changeDirectionState, hasSortOption } from './functions';
+import { PaginationOptions } from './interfaces';
+
+/**  Render sort options.*/
+export const renderSortOptions = (): void => {
+  const sortOptions = document.querySelectorAll('.sort');
+  const sortOption = document.querySelector<HTMLSelectElement>('#sort-option');
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/sort.ts:14]: Its value could be 'null', so we could not use Annotations here

File: apps/vanilla/src/scripts/sort.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+    });
+
+    SORT_DIRECTION.forEach(direction => {
+      sortDirection.innerHTML += `<option value="${direction.value}" class="sort-option">${direction.text}</option>`;
+    });
+    changeDirectionState(hasSortOption(SORT_SETTINGS.option));
+    sortOptions.forEach(element => {
+      element.addEventListener('change', async(): Promise<void> => {
+        PAGINATION_STATE.page = FIRST_PAGE;
+        PAGINATION_STATE.active = FIRST_PAGE;
```
Comment (thanhnguyen-saritasa) [inline apps/vanilla/src/scripts/sort.ts:32]: Seen SORT_SETTINGS has read-only property. so can not assign value here

File: apps/vanilla/src/scripts/functions.ts (hunk 13 above)
Comment (daoduong-saritasa) [reply apps/vanilla/src/scripts/functions.ts:22]: I have created a new utility function assert for null. Please check commit [6612870]

File: apps/vanilla/src/global/components.css [hunk 18]
```diff
@@ -1,28 +1,12 @@
...
-  border-bottom: 1px solid var(--link-color);
-}
-
-a:hover,
-:link:hover,
-.link:hover {
-  --link-color: var(--primary-light-color);
-}
-
 .links {
```
Comment (golovachev-yaroslav) [inline apps/vanilla/src/global/components.css:1]: Is it class for many links or for singular element?

File: apps/vanilla/src/global/index.css [hunk 19]
```diff
@@ -1,3 +1,5 @@
 @import url("@js-camp/theme/src/index.css");
 @import url("./base.css");
 @import url("./components.css");
+@import url("../../../.././node_modules/materialize-css/dist/css/materialize.min.css");
```
Comment (golovachev-yaroslav) [inline apps/vanilla/src/global/index.css:4]: Firstable, need to import external libs, otherwise it will override developer's style.

File: apps/vanilla/src/pages/index.css [hunk 20]
```diff
@@ -3,3 +3,21 @@
...
+}
+
+#pagination-wrapper {
```
Comment (golovachev-yaroslav) [inline apps/vanilla/src/pages/index.css:12]: Do not use id for styling. Id good for JS.
Also, id could not repeat on a page many times.

File: apps/vanilla/src/scripts/pagination.ts [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+  for (let page = maxLeft; page <= maxRight; page++) {
+    wrapper.innerHTML += `
+    <li index=${page} class="waves-effect pagination__item ${PAGINATION_STATE.active === page ? 'active' : ''}"><a href="#!">
+    ${page}
+      </a>
+    </li>`;
+  }
+  wrapper.innerHTML = `
+    <li index=${FIRST_PAGE} class="waves-effect pagination__item"><a>First</a></li>
```
Comment (golovachev-yaroslav) [inline apps/vanilla/src/scripts/pagination.ts:44]: Tag `a` requires `href` attribute, maybe replace it with another tag?

**Review by golovachev-yaroslav — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/global/components.css (hunk 18 above)
Comment (daoduong-saritasa) [reply apps/vanilla/src/global/components.css:3]: It's the default css in the boilerplate. I don't use it but I forgot to delete

File: apps/vanilla/src/scripts/anime.ts [hunk 22]
```diff
@@ -4,17 +4,16 @@ import { PaginationDto } from '@js-camp/core/dtos/pagination.dto';
...
-import { PaginationConfig } from './interfaces';
 
 /**
  * Get anime data from the server.
  * @param PaginationConfig Option for request parameters.
  */
 export const getAnime = async({ limit, page, ordering }: PaginationConfig): Promise<Pagination<Anime>> => {
-  const offset: number = limit * (page - 1);
-  const params = new URLSearchParams({ limit: limit.toString(), offset: offset.toString(), ordering });
+  const params = new URLSearchParams({ limit: limit.toString(), offset: (limit * (page - 1)).toString(), ordering });
```
Comment (golovachev-yaroslav) [inline apps/vanilla/src/scripts/anime.ts:16]: I think it would be better `(limit * (page - 1)).toString()` make as const.

**Review by golovachev-yaroslav — CHANGES_REQUESTED** [review]

File: libs/core/interfaces/pagination.ts [hunk 23]
```diff
@@ -0,0 +1,12 @@
...
+export interface PaginationConfig {
+
+  /** Number of items per request. */
+  readonly limit: number;
+
+  /**  */
+  readonly page: number;
+
+  /** Setting for sort features. */
+  readonly ordering: string;
```
Comment (panov-aleksey-saritasa) [inline libs/core/interfaces/pagination.ts:11]: It's not clear from the docs what the `ordering` property is.

File: apps/vanilla/src/scripts/sort.ts [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+      localStorage.setItem('active', FIRST_PAGE.toString());
+      const sortSetting = sortDirection.value + sortOption.value;
+      localStorage.setItem('sort', sortSetting);
+      const paginationConfig: PaginationConfig = {
+        limit: LIMIT,
+        page: FIRST_PAGE,
+        ordering: sortSetting,
+      };
+      setDirectionState(hasSortOption(sortOption.value));
+      const data: Pagination<Anime> = await getAnime(paginationConfig);
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/scripts/sort.ts:40]: Same about type annotations.

File: apps/vanilla/src/scripts/pagination.ts [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+import { LIMIT, FIRST_PAGE, HALF_NUMBER_OF_PAGES, NUMBER_OF_PAGES } from './variables';
+import { renderAnimeTable } from './animeTable';
+
+import { getAnime } from './anime';
+
+/**
+ * Render pagination.
+ * @param pages Number of total pages.
+ * */
+export const renderPaginateButton = (pages: number): void => {
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:15]: Prefer using [function declaration](https://wiki.saritasa.rocks/frontend/languages/ts-js/best-practices#use-function-declaration) instead

File: apps/vanilla/src/pages/index.css [hunk 26]
```diff
@@ -3,3 +3,29 @@
...
+  padding: 0 20px;
+}
+
+.pagination__item {
+  width: 3rem;
+}
+
+.active {
+  color: var(--white-color);
+}
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/pages/index.css:31]: If you are using [BEM](https://en.bem.info/), consider making use of [modifiers](https://en.bem.info/methodology/quick-start/#modifier). Right now it's quite implicit what this class is about, where it's going to be used

File: apps/vanilla/src/scripts/anime.ts (hunk 22 above)
Comment (panov-aleksey-saritasa) [reply apps/vanilla/src/scripts/anime.ts:16]: Consider making a mapper function that would convert `PaginationConfig` to the object with `limit` and `offset` properties

File: apps/vanilla/src/scripts/anime.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Get anime data from the server.
+ * @param PaginationConfig Option for request parameters.
+ */
+export const getAnime = async({ limit, page, ordering }: PaginationConfig): Promise<Pagination<Anime>> => {
+  const params = new URLSearchParams({ limit: limit.toString(), offset: (limit * (page - 1)).toString(), ordering });
+  const { data } = await api.get<PaginationDto<AnimeDto>>(
+    `anime/anime/`, { params },
+  );
+  return PaginationMapper.fromDto<AnimeDto, Anime>(data, animeDto => AnimeMapper.fromDto(animeDto));
```
Comment (panov-aleksey-saritasa) [inline apps/vanilla/src/scripts/anime.ts:20]:
```suggestion
  return PaginationMapper.fromDto(data, animeDto => AnimeMapper.fromDto(animeDto));
```

**Review by panov-aleksey-saritasa — CHANGES_REQUESTED** [review]

**Review by elena-dragunova — APPROVED** [review]

File: apps/vanilla/src/global/variables.css [hunk 28]
```diff
@@ -4,4 +4,7 @@
   --heading-color-1: gold;
   --heading-color-2: darksalmon;
   --heading-color-3: powderblue;
+  --default-layout-width: 90%;
+  --white-color: #fff;
```
Comment (TienLuu-Saritasa) [inline apps/vanilla/src/global/variables.css:8]: Please rename your variable
https://wiki.saritasa.rocks/frontend/languages/css/#naming

**Review by TienLuu-Saritasa — CHANGES_REQUESTED** [review]

File: apps/vanilla/src/global/variables.css [hunk 29]
```diff
@@ -1,7 +1,10 @@
...
    * you shouldn't be using color names in real projects. */
-  --heading-color-1: gold;
-  --heading-color-2: darksalmon;
-  --heading-color-3: powderblue;
+  --heading-color-1: #ffb57a;
+  --heading-color-2: #f16a00;
+  --heading-color-3: #7bf8ff;
+  --default-layout-width: 90%;
+  --active-background-color: #ff8525;
+  --active-text-color: #fff;
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/global/variables.css:9]: Please group your variables

File: apps/vanilla/src/pages/index.html [hunk 30]
```diff
@@ -4,12 +4,28 @@
...
     <h1 class="title">Main page ⚡</h1>
-    <div class="links">
-      <a href="/example/">Go to example page</a>
+
+    <ul class="pagination">
+
+    </ul>
+      <!-- Select section -->
+    <div class="sort">
+      <div class="sort__select-element">
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/pages/index.html:17]: You should use label instead of just div(class="sort__select-element"). That way "Sort by" would be associated with select, that improves accessibility.

File: apps/vanilla/src/scripts/animeTable.ts [hunk 31]
```diff
@@ -0,0 +1,49 @@
...
+    <tr>
+         <td>
+           <img class="anime-image" src="${currentValue.image}" alt="Anime image">
+         </td>
+         <td>${currentValue.englishTitle}</td>
+         <td>${currentValue.japaneseTitle}</td>
+         <td>${dateConverter(currentValue.aired.start)}</td>
+         <td>${currentValue.type}</td>
+         <td>${currentValue.status}</td>
+       </tr>
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/animeTable.ts:27]: Please fix indentation

File: apps/vanilla/src/scripts/pagination.ts [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+      const sortSetting = localStorage.getItem('sort');
+      assertNonNullish(sortSetting);
+      const paginationConfig: PaginationConfig = {
+        limit: LIMIT,
+        page: parseInt(currentIndex, 10),
+        ordering: sortSetting,
+      };
+      const data = await getAnime(paginationConfig);
+      renderAnimeTable(data);
+    });
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:63]: Please group you code lines, it is hard to read.

**Review by TienLuu-Saritasa — APPROVED** [review]

File: apps/vanilla/src/scripts/pagination.ts [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+    <li index=${page} class="waves-effect pagination__item ${page === currentPage ? 'pagination__item-active' : ''}">
+    ${page}
+    </li>`;
+  }
+  wrapper.innerHTML = `
+    <li index=${FIRST_PAGE} class="waves-effect pagination__item">First</li>
+    ${wrapper.innerHTML}`;
+
+  wrapper.innerHTML += `<li index=${pages} class="waves-effect pagination__item">Last</li>`;
+  const paginateButton = document.querySelectorAll('.pagination__item');
```
Comment (artem-sotnichenko-saritasa) [inline apps/vanilla/src/scripts/pagination.ts:48]: You should use button for interactivity, `li` is not designed for this purpose. You could just put `button` inside `li`, or don't use `ul` here, because pagination is not a list.

Comment (daoduong-saritasa) [reply apps/vanilla/src/scripts/pagination.ts:48]: It's materialize pagination. Should I rewrite it?

File: apps/vanilla/src/scripts/pagination.ts [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+
+import { LIMIT, FIRST_PAGE, HALF_NUMBER_OF_PAGES, NUMBER_OF_PAGES } from './variables';
+import { renderAnimeTable } from './animeTable';
+
+import { getAnime } from './anime';
+
+/**
+ * Render pagination.
+ * @param pages Number of total pages.
+ * */
```
Comment (chernodub) [inline apps/vanilla/src/scripts/pagination.ts:12]: Pls make sure your code is formatted properly

File: apps/vanilla/src/scripts/API.ts [hunk 35]
```diff
@@ -0,0 +1,6 @@
+import axios from 'axios';
```
Comment (chernodub) [inline apps/vanilla/src/scripts/API.ts:1]: Usually we prefer not to uppercase names of the files, even if abbreviations are used, please stick to that

File: apps/vanilla/src/scripts/pagination.ts [hunk 36]
```diff
@@ -0,0 +1,72 @@
...
+export function renderPaginateButton(pages: number): void {
```
Comment (chernodub) [inline apps/vanilla/src/scripts/pagination.ts:13]: As far as I understand, this function not only renders the whole pagination controller, instead of a single button. Please keep your naming and implementation consistent

File: apps/vanilla/src/scripts/pagination.ts [hunk 37]
```diff
@@ -0,0 +1,72 @@
...
+  const pageValueFromStorage = localStorage.getItem('active');
+  assertNonNullish(pageValueFromStorage);
+  const currentPage = parseInt(pageValueFromStorage, 10);
+  const wrapper = document.querySelector<HTMLDivElement>('.pagination');
+  assertNonNullish(wrapper);
+  wrapper.innerHTML = ``;
+  let maxLeft = (currentPage - HALF_NUMBER_OF_PAGES);
+  let maxRight = (currentPage + HALF_NUMBER_OF_PAGES);
```
Comment (chernodub) [inline apps/vanilla/src/scripts/pagination.ts:21]: Please revise the naming of these variables. Think from a perspective of a code reader, I see the function `renderPaginationButton`, then I see `maxRight` and `maxLeft` inside. This logic is very hard to follow. 

After you wrote a function, consider re-reading it like a small narrative, try to keep your plot consecutive. You are a narrator in a way :)

File: apps/vanilla/src/scripts/pagination.ts (hunk 33 above)
Comment (chernodub) [reply apps/vanilla/src/scripts/pagination.ts:48]: [@daoduong-saritasa you should always use interactive elements for handling user events](https://wiki.saritasa.rocks/frontend/languages/html#use-button-for-handling-user-clicks). It is not clear what do you mean by `materialize pagination`? You apply an event listener to a `li` element, this is the problem

File: apps/vanilla/src/scripts/search.ts [hunk 38]
```diff
@@ -0,0 +1,30 @@
...
+import { renderAnimeTable } from './animeTable';
+
+import { FIRST_PAGE, LIMIT } from './variables';
+
+/** Search feature. */
+export function performSearching(): void {
+  const searchInputElement = document.querySelector<HTMLInputElement>('.search_input');
+  const searchButtonElement = document.querySelector<HTMLButtonElement>('.search_button');
+  assertNonNullish(searchInputElement);
+  assertNonNullish(searchButtonElement);
```
Comment (chernodub) [inline apps/vanilla/src/scripts/search.ts:14]: The properties above are not nullable, right? If so, you don't need these assertions

File: apps/vanilla/src/scripts/search.ts [hunk 39]
```diff
@@ -0,0 +1,30 @@
+import { PaginationConfig } from '@js-camp/core/interfaces/pagination';
+import { assertNonNullish } from '@js-camp/core/utils/assertNonNullish';
+
+import { getAnime } from './anime';
+import { renderAnimeTable } from './animeTable';
+
+import { FIRST_PAGE, LIMIT } from './variables';
+
+/** Search feature. */
+export function performSearching(): void {
```
Comment (chernodub) [inline apps/vanilla/src/scripts/search.ts:10]: As far as I see, this function initializes search elements with event listeners. Consider revising the name, please.

File: apps/vanilla/src/scripts/sort.ts [hunk 40]
```diff
@@ -0,0 +1,50 @@
+import { assertNonNullish } from '@js-camp/core/utils/assertNonNullish';
+import { PaginationConfig } from '@js-camp/core/interfaces/pagination';
+
+import { renderAnimeTable } from './animeTable';
+import { SORT_DIRECTIONS, SORT_OPTIONS, LIMIT, FIRST_PAGE } from './variables';
+import { getAnime } from './anime';
+import { setDirectionState, hasSortOption } from './functions';
+
+/**  Render sort options.*/
```
Comment (chernodub) [inline apps/vanilla/src/scripts/sort.ts:9]: Please keep your formatting consistent for every JSDoc
```
/** Render sort options. */
```

File: libs/core/dtos/anime.dto.ts [hunk 41]
```diff
@@ -0,0 +1,50 @@
+/** Aired DTO. */
```
Comment (chernodub) [inline libs/core/dtos/anime.dto.ts:1]: Why is this comment needed?

File: libs/core/dtos/anime.dto.ts [hunk 42]
```diff
@@ -0,0 +1,50 @@
...
+
+/** Possible values for type. */
+export enum Type {
+  'TV',
+  'OVA',
+  'MOVIE',
+  'SPECIAL',
+  'ONA',
+  'MUSIC',
+}
```
Comment (chernodub) [inline libs/core/dtos/anime.dto.ts:12]: Seems that you've used `enum` wrong. Checkout [this demo](https://www.typescriptlang.org/play?#code/KYOwrgtgBAKgngB2FA3gKClA5DAalgGg2wHlcBBQ4rAWTIEkBRKzLAZQAVGBhe8gGRakAcpSKsaAVTb1uVAL5o0AYwD2IAM6qANsAB021QHMAFPCR68ASiA), click `run` and see the log. The real values for these properties are numbers. But BE sends you strings. So, for your case you could use a [string union alias](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html). 

Please revise [enums article](https://www.typescriptlang.org/docs/handbook/enums.html#handbook-content)

File: apps/vanilla/src/scripts/search.ts (hunk 38 above)
Comment (daoduong-saritasa) [reply apps/vanilla/src/scripts/search.ts:14]: The querySelector always return the element or null so it's nullable

File: apps/vanilla/src/scripts/pagination.ts [hunk 43]
```diff
@@ -0,0 +1,72 @@
...
+ * @param pages Number of total pages.
+ * */
+export function renderPaginateButton(pages: number): void {
+  const pageValueFromStorage = localStorage.getItem('active');
+  assertNonNullish(pageValueFromStorage);
+  const currentPage = parseInt(pageValueFromStorage, 10);
+  const wrapper = document.querySelector<HTMLDivElement>('.pagination');
+  assertNonNullish(wrapper);
+  wrapper.innerHTML = ``;
+  let maxLeft = (currentPage - HALF_NUMBER_OF_PAGES);
```
[the remark this answers is not in the document]
Comment (daoduong-saritasa) [reply apps/vanilla/src/scripts/pagination.ts]: I don't understand about "make it global" part. Here I use const and let for declaration, I thought it declare a block-scoped variable and I declare it in a function.

File: apps/vanilla/src/global/index.css [hunk 44]
```diff
@@ -1,3 +1,4 @@
...
```
[the remark this answers is not in the document]
Comment (daoduong-saritasa) [reply apps/vanilla/src/global/index.css:1]: When I remove the dots I get this error. I search and saw a PR that fix this error but I still get this
image

File: apps/vanilla/src/scripts/pagination.ts (hunk 43 above)
[the remark this answers is not in the document]
Comment (chernodub) [reply apps/vanilla/src/scripts/pagination.ts]: @daoduong-saritasa I'm talking about `HALF_NUMBER_OF_PAGES` constant. It is only used in this file, so there is not point making it globally available for the whole application

File: apps/vanilla/src/scripts/animeTable.ts [hunk 45]
```diff
@@ -0,0 +1,48 @@
...
+import { LIMIT } from './variables';
+
+/**
+ * Render anime table.
+ * @param dataset Data for render.
+ */
+export function renderAnimeTable(dataset: Pagination<Anime>): void {
+  const dataTable = document.querySelector<HTMLDivElement>('.anime-table');
+  assertNonNullish(dataTable);
+  const tableBody = dataset.results.reduce((previousValue: string, currentValue: Anime): string => `${previousValue}
```
Comment (pavelzadorin-saritasa) [inline apps/vanilla/src/scripts/animeTable.ts:17]: and again about the names, compare, let's say somewhere in another function you see the same line, but we don't use Typescript:
```const tableBody = dataset.results.reduce((previousValue, currentValue)=> `${previousValue}<tr><td>${currentValue.name}</td></tr>```
Will it be easy clear from your code what it's doing? 
or you see this:
```const tableBody = studioList.results.reduce((previousBody, studio) => `${previousBody}<tr><td>${studio.name}</td></tr>```
It seems to me that the second one is obvious, even without Typescript, but it is impossible to understand the first one without analyzing where our function is called and what will be passed to it in the parameter `dataset`.

Comment (pavelzadorin-saritasa) [inline apps/vanilla/src/scripts/animeTable.ts:17]: and second thing, since the parameters `previousValue` and `currentValue` use the same term `Value`, then it is logical to expect that this is the previous and current values of the same thing, right? well, i.e. for example the previous Anime and the current Anime. But you have completely different entities, the first is the accum of the table body, and the second is Anime, which means it is desirable to reflect this in the variable names so as not to confuse other developers.

File: libs/core/models/dateRange.ts [hunk 46]
```diff
@@ -0,0 +1,18 @@
+import { OmitImmerable, Immerable } from './immerable';
+
+/** Anime aired time. */
+export class DateRange extends Immerable {
+
+  /** Anime start date. */
+  public readonly start: string;
```
Comment (pavelzadorin-saritasa) [inline libs/core/models/dateRange.ts:7]: If this is a date, then it is logical to store it in the Date type, because it is much more convenient to work with it in the future. For example, you don’t have to pass a string to the convertDate function and convert it to a date there in order to get the string again, but in a different format ...
This is one of the reasons why we separate the model layer from the repository layer. A model is a data structure that is convenient for us, for our application, for working in JS. And it doesn’t matter that the data comes from the API in a different format (after all, only simple data structures are allowed in JSON), for this there are Mappers that convert raw / inconvenient data into a format convenient for us.

File: apps/vanilla/src/scripts/sort.ts [hunk 47]
```diff
@@ -0,0 +1,48 @@
...
+      const sortSetting = sortDirection.value + sortOption.value;
+      localStorage.setItem('sort', sortSetting);
+
+      setDirectionState(hasSortOption(sortOption.value));
+      const searchQuery = localStorage.getItem('search');
+      assertNonNullish(searchQuery);
+
+      const paginationConfig: PaginationConfig = {
+        limit: LIMIT,
+        page: FIRST_PAGE,
```
Comment (pavelzadorin-saritasa) [inline apps/vanilla/src/scripts/sort.ts:39]: it turns out when you sort your page is reset to the first? I do not think that this is justified, the number of pages will not change from using sorting and we may well leave the user on the current open page, this will be a good UX.
resetting pagination is justified for filtering and searching, because the number of pages changes, but not for sorting.
for example: https://mui.com/material-ui/react-table/#data-table

File: apps/vanilla/src/pages/index.ts [hunk 48]
```diff
@@ -0,0 +1,29 @@
...
+  localStorage.setItem('sort', '');
+  localStorage.setItem('search', '');
+  const paginationConfig: PaginationConfig = {
+    limit: LIMIT,
+    page: FIRST_PAGE,
+    ordering: '',
+    search: '',
+  };
+
+  const anime = await getAnime(paginationConfig);
```
Comment (pavelzadorin-saritasa) [inline apps/vanilla/src/pages/index.ts:25]: `anime` is a singular noun (1 film), but you get a list of anime, so you have to use plurar here, better naming would be `animes` or `animeList`

File: apps/vanilla/src/scripts/animeTable.ts [hunk 49]
```diff
@@ -0,0 +1,48 @@
...
+import { LIMIT } from './variables';
+
+/**
+ * Render anime table.
+ * @param dataset Data for rendering.
+ */
+export function renderTable(dataset: Pagination<Anime>): void {
+  const dataTable = document.querySelector<HTMLDivElement>('.anime-table');
+  assertNonNullish(dataTable);
+  const tableBody = dataset.results.reduce((previousBody: string, currentData: Anime): string => `${previousBody}
```
Comment (pavelzadorin-saritasa) [inline apps/vanilla/src/scripts/animeTable.ts:17]: `dataset` and `currentData` are not semantic names, please take a closer look at my previous [comment](https://github.com/saritasa-nest/camp-js-2022-dao-duong/pull/1#discussion_r924766602), I'm sure the option I proposed will improve the readability of the code.

**Review by pavelzadorin-saritasa — APPROVED** [review]

**Review by artem-sotnichenko-saritasa — APPROVED** [review]

**Review by chernodub — APPROVED** [review]
